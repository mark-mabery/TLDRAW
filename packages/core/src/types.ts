/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* --------------------- Primary -------------------- */

import type React from 'react'

export type Patch<T> = Partial<{ [P in keyof T]: T | Partial<T> | Patch<T[P]> }>

type ForwardedRef<T> = ((instance: T | null) => void) | React.MutableRefObject<T | null> | null

export interface TLPage<T extends TLShape, B extends TLBinding> {
  id: string
  name?: string
  childIndex?: number
  shapes: Record<string, T>
  bindings: Record<string, B>
}

export interface TLPageState {
  id: string
  selectedIds: string[]
  camera: {
    point: number[]
    zoom: number
  }
  brush?: TLBounds | null
  pointedId?: string | null
  hoveredId?: string | null
  editingId?: string | null
  bindingId?: string | null
  boundsRotation?: number
  currentParentId?: string | null
}

export interface TLHandle {
  id: string
  index: number
  point: number[]
  canBind?: boolean
  bindingId?: string
}

export interface TLShape {
  id: string
  type: string
  parentId: string
  childIndex: number
  name: string
  point: number[]
  rotation?: number
  children?: string[]
  handles?: Record<string, TLHandle>
  isLocked?: boolean
  isHidden?: boolean
  isEditing?: boolean
  isGenerated?: boolean
  isAspectRatioLocked?: boolean
}

export type TLShapeUtils<
  T extends TLShape = any,
  E extends Element = any,
  M = any,
  K = any
> = Record<string, TLShapeUtil<T, E, M, K>>

export interface TLRenderInfo<T extends TLShape, E = any, M = any> {
  shape: T
  isEditing: boolean
  isBinding: boolean
  isHovered: boolean
  isSelected: boolean
  isCurrentParent: boolean
  meta: M extends any ? M : never
  onShapeChange?: TLCallbacks<T>['onShapeChange']
  onShapeBlur?: TLCallbacks<T>['onShapeBlur']
  events: {
    onPointerDown: (e: React.PointerEvent<E>) => void
    onPointerUp: (e: React.PointerEvent<E>) => void
    onPointerEnter: (e: React.PointerEvent<E>) => void
    onPointerMove: (e: React.PointerEvent<E>) => void
    onPointerLeave: (e: React.PointerEvent<E>) => void
  }
}

export interface TLShapeProps<T extends TLShape, E = any, M = any> extends TLRenderInfo<T, E, M> {
  ref: ForwardedRef<E>
  shape: T
}

export interface TLTool {
  id: string
  name: string
}

export interface TLBinding<M = any> {
  id: string
  type: string
  toId: string
  fromId: string
  meta: M
}

export interface TLTheme {
  brushFill?: string
  brushStroke?: string
  selectFill?: string
  selectStroke?: string
  background?: string
  foreground?: string
}

export type TLWheelEventHandler = (
  info: TLPointerInfo<string>,
  e: React.WheelEvent<Element> | WheelEvent
) => void

export type TLPinchEventHandler = (
  info: TLPointerInfo<string>,
  e:
    | React.WheelEvent<Element>
    | WheelEvent
    | React.TouchEvent<Element>
    | TouchEvent
    | React.PointerEvent<Element>
    | PointerEventInit
) => void

export type TLShapeChangeHandler<T, K = any> = (
  shape: { id: string } & Partial<T>,
  info?: K
) => void

export type TLShapeBlurHandler<K = any> = (info?: K) => void

export type TLPointerEventHandler = (info: TLPointerInfo<string>, e: React.PointerEvent) => void

export type TLCanvasEventHandler = (info: TLPointerInfo<'canvas'>, e: React.PointerEvent) => void

export type TLBoundsEventHandler = (info: TLPointerInfo<'bounds'>, e: React.PointerEvent) => void

export type TLBoundsHandleEventHandler = (
  info: TLPointerInfo<TLBoundsCorner | TLBoundsEdge | 'rotate'>,
  e: React.PointerEvent
) => void

export interface TLCallbacks<T extends TLShape> {
  // Camera events
  onPinchStart: TLPinchEventHandler
  onPinchEnd: TLPinchEventHandler
  onPinch: TLPinchEventHandler
  onPan: TLWheelEventHandler
  onZoom: TLWheelEventHandler

  // Pointer Events
  onPointerMove: TLPointerEventHandler
  onPointerUp: TLPointerEventHandler
  onPointerDown: TLPointerEventHandler

  // Canvas (background)
  onPointCanvas: TLCanvasEventHandler
  onDoubleClickCanvas: TLCanvasEventHandler
  onRightPointCanvas: TLCanvasEventHandler
  onDragCanvas: TLCanvasEventHandler
  onReleaseCanvas: TLCanvasEventHandler

  // Shape
  onPointShape: TLPointerEventHandler
  onDoubleClickShape: TLPointerEventHandler
  onRightPointShape: TLPointerEventHandler
  onDragShape: TLPointerEventHandler
  onHoverShape: TLPointerEventHandler
  onUnhoverShape: TLPointerEventHandler
  onReleaseShape: TLPointerEventHandler

  // Bounds (bounding box background)
  onPointBounds: TLBoundsEventHandler
  onDoubleClickBounds: TLBoundsEventHandler
  onRightPointBounds: TLBoundsEventHandler
  onDragBounds: TLBoundsEventHandler
  onHoverBounds: TLBoundsEventHandler
  onUnhoverBounds: TLBoundsEventHandler
  onReleaseBounds: TLBoundsEventHandler

  // Bounds handles (corners, edges)
  onPointBoundsHandle: TLBoundsHandleEventHandler
  onDoubleClickBoundsHandle: TLBoundsHandleEventHandler
  onRightPointBoundsHandle: TLBoundsHandleEventHandler
  onDragBoundsHandle: TLBoundsHandleEventHandler
  onHoverBoundsHandle: TLBoundsHandleEventHandler
  onUnhoverBoundsHandle: TLBoundsHandleEventHandler
  onReleaseBoundsHandle: TLBoundsHandleEventHandler

  // Handles (ie the handles of a selected arrow)
  onPointHandle: TLPointerEventHandler
  onDoubleClickHandle: TLPointerEventHandler
  onRightPointHandle: TLPointerEventHandler
  onDragHandle: TLPointerEventHandler
  onHoverHandle: TLPointerEventHandler
  onUnhoverHandle: TLPointerEventHandler
  onReleaseHandle: TLPointerEventHandler

  // Misc
  onShapeChange: TLShapeChangeHandler<T, any>
  onShapeBlur: TLShapeBlurHandler<any>
  onRenderCountChange: (ids: string[]) => void
  onError: (error: Error) => void
  onBoundsChange: (bounds: TLBounds) => void
}

export interface TLBounds {
  minX: number
  minY: number
  maxX: number
  maxY: number
  width: number
  height: number
  rotation?: number
}

export type TLIntersection = {
  didIntersect: boolean
  message: string
  points: number[][]
}

export enum TLBoundsEdge {
  Top = 'top_edge',
  Right = 'right_edge',
  Bottom = 'bottom_edge',
  Left = 'left_edge',
}

export enum TLBoundsCorner {
  TopLeft = 'top_left_corner',
  TopRight = 'top_right_corner',
  BottomRight = 'bottom_right_corner',
  BottomLeft = 'bottom_left_corner',
}

export interface TLPointerInfo<T extends string = string> {
  target: T
  pointerId: number
  origin: number[]
  point: number[]
  delta: number[]
  pressure: number
  shiftKey: boolean
  ctrlKey: boolean
  metaKey: boolean
  altKey: boolean
}

export interface TLKeyboardInfo {
  origin: number[]
  point: number[]
  key: string
  keys: string[]
  shiftKey: boolean
  ctrlKey: boolean
  metaKey: boolean
  altKey: boolean
}

export interface TLTransformInfo<T extends TLShape> {
  type: TLBoundsEdge | TLBoundsCorner
  initialShape: T
  scaleX: number
  scaleY: number
  transformOrigin: number[]
}

export interface TLBezierCurveSegment {
  start: number[]
  tangentStart: number[]
  normalStart: number[]
  pressureStart: number
  end: number[]
  tangentEnd: number[]
  normalEnd: number[]
  pressureEnd: number
}

/* -------------------------------------------------- */
/*                   Shape Utility                    */
/* -------------------------------------------------- */

export type TLShapeUtil<
  T extends TLShape,
  E extends Element,
  M = any,
  K = { [key: string]: any }
> = K & {
  type: T['type']

  defaultProps: T

  Component(
    this: TLShapeUtil<T, E, M>,
    props: TLRenderInfo<T, E, M>,
    ref: ForwardedRef<E>
  ): React.ReactElement<TLRenderInfo<T, E, M>, E['tagName']>

  Indicator(
    this: TLShapeUtil<T, E, M>,
    props: { shape: T; meta: M; isHovered: boolean; isSelected: boolean }
  ): React.ReactElement | null

  getBounds(this: TLShapeUtil<T, E, M>, shape: T): TLBounds

  refMap: Map<string, React.RefObject<E>>

  boundsCache: WeakMap<TLShape, TLBounds>

  isAspectRatioLocked: boolean

  canEdit: boolean

  canBind: boolean

  getRotatedBounds(this: TLShapeUtil<T, E, M>, shape: T): TLBounds

  hitTest(this: TLShapeUtil<T, E, M>, shape: T, point: number[]): boolean

  hitTestBounds(this: TLShapeUtil<T, E, M>, shape: T, bounds: TLBounds): boolean

  shouldRender(this: TLShapeUtil<T, E, M>, prev: T, next: T): boolean

  getCenter(this: TLShapeUtil<T, E, M>, shape: T): number[]

  getRef(this: TLShapeUtil<T, E, M>, shape: T): React.RefObject<E>

  getBindingPoint<K extends TLShape>(
    this: TLShapeUtil<T, E, M>,
    shape: T,
    fromShape: K,
    point: number[],
    origin: number[],
    direction: number[],
    padding: number,
    bindAnywhere: boolean
  ): { point: number[]; distance: number } | undefined

  create: (this: TLShapeUtil<T, E, M>, props: { id: string } & Partial<T>) => T

  mutate: (this: TLShapeUtil<T, E, M>, shape: T, props: Partial<T>) => Partial<T>

  transform: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    bounds: TLBounds,
    info: TLTransformInfo<T>
  ) => Partial<T> | void

  transformSingle: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    bounds: TLBounds,
    info: TLTransformInfo<T>
  ) => Partial<T> | void

  updateChildren: <K extends TLShape>(
    this: TLShapeUtil<T, E, M>,
    shape: T,
    children: K[]
  ) => Partial<K>[] | void

  onChildrenChange: (this: TLShapeUtil<T, E, M>, shape: T, children: TLShape[]) => Partial<T> | void

  onBindingChange: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    binding: TLBinding,
    target: TLShape,
    targetBounds: TLBounds,
    center: number[]
  ) => Partial<T> | void

  onHandleChange: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    handle: Partial<T['handles']>,
    info: Partial<TLPointerInfo>
  ) => Partial<T> | void

  onRightPointHandle: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    handle: Partial<T['handles']>,
    info: Partial<TLPointerInfo>
  ) => Partial<T> | void

  onDoubleClickHandle: (
    this: TLShapeUtil<T, E, M>,
    shape: T,
    handle: Partial<T['handles']>,
    info: Partial<TLPointerInfo>
  ) => Partial<T> | void

  onDoubleClickBoundsHandle: (this: TLShapeUtil<T, E, M>, shape: T) => Partial<T> | void

  onSessionComplete: (this: TLShapeUtil<T, E, M>, shape: T) => Partial<T> | void

  onStyleChange: (this: TLShapeUtil<T, E, M>, shape: T) => Partial<T> | void

  _Component: React.ForwardRefExoticComponent<any>
}

// export interface TLShapeUtil<T extends TLShape, E extends Element, M = any>
//   extends TLShapeUtilRequired<T, E, M>,
//     Required<TLShapeUtilDefaults<T, E>> {
//   _Component: React.ForwardRefExoticComponent<any> & {
//     defaultProps: any
//     propTypes: any
//   }
// }

// export interface TLShapeUtilConfig<T extends TLShape, E extends Element, M = any>
//   extends TLShapeUtilRequired<T, E, M>,
//     Partial<TLShapeUtilDefaults<T, E>> {}

/* -------------------- Internal -------------------- */

export interface IShapeTreeNode<T extends TLShape, M = any> {
  shape: T
  children?: IShapeTreeNode<TLShape, M>[]
  isEditing: boolean
  isBinding: boolean
  isHovered: boolean
  isSelected: boolean
  isCurrentParent: boolean
  meta?: M extends any ? M : never
}

/* -------------------------------------------------- */
/*                    Utility Types                   */
/* -------------------------------------------------- */

export type MappedByType<K extends string, T extends { type: K }> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [P in T['type']]: T extends any ? (P extends T['type'] ? T : never) : never
}

export type RequiredKeys<T> = {
  [K in keyof T]-?: Record<string, unknown> extends Pick<T, K> ? never : K
}[keyof T]
