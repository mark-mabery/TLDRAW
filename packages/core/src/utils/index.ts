import { Utils } from './utils'
export { Utils } from './utils'
export { Svg } from './svg'

export default Utils
