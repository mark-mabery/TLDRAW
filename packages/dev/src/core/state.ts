import type {
  TLBinding,
  TLPage,
  TLPageState,
  TLPointerEventHandler,
  TLShapeChangeHandler,
} from '@tldraw/core'
import type { RectangleShape } from './box'
import type { LabelShape } from './label'
import { StateManager } from 'rko'

type Shapes = RectangleShape | LabelShape

interface State {
  page: TLPage<Shapes, TLBinding>
  pageState: TLPageState
  meta: {
    isDarkMode: boolean
  }
}

class AppState extends StateManager<State> {
  /* ----------------------- API ---------------------- */

  selectShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
      },
    })
  }

  deselect() {
    this.patchState({
      pageState: {
        selectedIds: [],
        editingId: undefined,
      },
    })
  }

  startEditingShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
        editingId: shapeId,
      },
    })
  }

  changeShapeText = (id: string, text: string) => {
    this.patchState({
      page: {
        shapes: {
          [id]: { text },
        },
      },
    })
  }

  /* --------------------- Events --------------------- */

  onPointCanvas: TLPointerEventHandler = (info) => {
    this.deselect()
  }

  onPointShape: TLPointerEventHandler = (info) => {
    this.selectShape(info.target)
  }

  onDoubleClickShape: TLPointerEventHandler = (info) => {
    this.startEditingShape(info.target)
  }

  onDoubleClickBounds: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerDown: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerUp: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerMove: TLPointerEventHandler = (info) => {
    // Todo
  }

  onShapeChange: TLShapeChangeHandler<Shapes> = (shape) => {
    if (shape.type === 'rectangle' && shape.size) {
      this.patchState({
        page: {
          shapes: {
            [shape.id]: { ...shape, size: [...shape.size] },
          },
        },
      })
    }
  }
}

export const appState = new AppState({
  page: {
    id: 'page1',
    shapes: {
      rect1: {
        id: 'rect1',
        parentId: 'page1',
        name: 'Rectangle',
        childIndex: 1,
        type: 'rectangle',
        point: [0, 0],
        rotation: 0,
        size: [100, 100],
        text: 'Hello world!',
      },
      label1: {
        id: 'label1',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [-200, -200],
        rotation: 0,
        text: 'My shape is stateful, I should still render while off screen!',
      },
      label2: {
        id: 'label2',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [200, 200],
        rotation: 0,
        text: 'Hello world!',
      },
    },
    bindings: {},
  },
  pageState: {
    id: 'page1',
    selectedIds: [],
    camera: {
      point: [0, 0],
      zoom: 1,
    },
  },
  meta: {
    isDarkMode: false,
  },
})
