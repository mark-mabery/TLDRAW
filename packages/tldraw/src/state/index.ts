export * from './tlstate'
export * from './state-types'
