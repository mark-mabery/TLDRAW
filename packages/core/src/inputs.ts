import type React from 'react'
import type { TLKeyboardInfo, TLPointerInfo } from './types'
import { Vec, Utils } from './utils'

const DOUBLE_CLICK_DURATION = 250

export class Inputs {
  pointer?: TLPointerInfo<string>
  keyboard?: TLKeyboardInfo
  keys: Record<string, boolean> = {}
  isPinching = false

  offset = [0, 0]
  size = [10, 10]

  pointerUpTime = 0

  activePointer?: number

  pointerIsValid(e: TouchEvent | React.TouchEvent | PointerEvent | React.PointerEvent) {
    if ('pointerId' in e) {
      if (this.activePointer && this.activePointer !== e.pointerId) return false
    }

    if ('touches' in e) {
      const touch = e.changedTouches[0]
      if (this.activePointer && this.activePointer !== touch.identifier) return false
    }

    return true
  }

  touchStart<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    this.activePointer = touch.identifier

    const info: TLPointerInfo<T> = {
      target,
      pointerId: touch.identifier,
      origin: Inputs.getPoint(touch),
      delta: [0, 0],
      point: Inputs.getPoint(touch),
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  touchEnd<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    const info: TLPointerInfo<T> = {
      target,
      pointerId: touch.identifier,
      origin: Inputs.getPoint(touch),
      delta: [0, 0],
      point: Inputs.getPoint(touch),
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    this.activePointer = undefined

    return info
  }

  touchMove<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    const prev = this.pointer

    const point = Inputs.getPoint(touch)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: touch.identifier,
      point,
      delta,
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerDown<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const point = Inputs.getPoint(e, this.offset)

    this.activePointer = e.pointerId

    const info: TLPointerInfo<T> = {
      target,
      pointerId: e.pointerId,
      origin: point,
      point: point,
      delta: [0, 0],
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerEnter<T extends string>(
    e: PointerEvent | React.PointerEvent,
    target: T
  ): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const point = Inputs.getPoint(e, this.offset)

    const info: TLPointerInfo<T> = {
      target,
      pointerId: e.pointerId,
      origin: point,
      delta: [0, 0],
      point: point,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerMove<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: e.pointerId,
      point,
      delta,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerUp<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    this.activePointer = undefined

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: e.pointerId,
      point,
      delta,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    this.pointerUpTime = Date.now()

    return info
  }

  panStart = (e: WheelEvent): TLPointerInfo<'wheel'> => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const info: TLPointerInfo<'wheel'> = {
      target: 'wheel',
      pointerId: this.pointer?.pointerId || 0,
      origin: this.pointer?.origin || [0, 0],
      delta: [0, 0],
      pressure: 0.5,
      point: Inputs.getPoint(e, this.offset),
      shiftKey,
      ctrlKey,
      metaKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pan = (delta: number[], e: WheelEvent): TLPointerInfo<'wheel'> => {
    if (!this.pointer || this.pointer.target !== 'wheel') {
      return this.panStart(e)
    }

    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const info: TLPointerInfo<'wheel'> = {
      ...prev,
      target: 'wheel',
      delta,
      point,
      shiftKey,
      ctrlKey,
      metaKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  isDoubleClick() {
    if (!this.pointer) return false

    const { origin, point } = this.pointer

    return Date.now() - this.pointerUpTime < DOUBLE_CLICK_DURATION && Vec.dist(origin, point) < 4
  }

  clear() {
    this.pointer = undefined
  }

  resetDoubleClick() {
    this.pointerUpTime = 0
  }

  keydown = (e: KeyboardEvent | React.KeyboardEvent): TLKeyboardInfo => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    this.keys[e.key] = true

    return {
      point: this.pointer?.point || [0, 0],
      origin: this.pointer?.origin || [0, 0],
      key: e.key,
      keys: Object.keys(this.keys),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }
  }

  keyup = (e: KeyboardEvent | React.KeyboardEvent): TLKeyboardInfo => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    delete this.keys[e.key]

    return {
      point: this.pointer?.point || [0, 0],
      origin: this.pointer?.origin || [0, 0],
      key: e.key,
      keys: Object.keys(this.keys),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }
  }

  pinch(point: number[], origin: number[]) {
    const { shiftKey, ctrlKey, metaKey, altKey } = this.keys

    const delta = Vec.sub(origin, point)

    const info: TLPointerInfo<'pinch'> = {
      pointerId: 0,
      target: 'pinch',
      origin,
      delta: delta,
      point: Vec.sub(Vec.round(point), this.offset),
      pressure: 0.5,
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  reset() {
    this.pointerUpTime = 0
    this.pointer = undefined
    this.keyboard = undefined
    this.activePointer = undefined
    this.keys = {}
  }

  static getPoint(
    e: PointerEvent | React.PointerEvent | Touch | React.Touch | WheelEvent,
    offset = [0, 0]
  ): number[] {
    return [
      Number(e.clientX.toPrecision(5)) - offset[0],
      Number(e.clientY.toPrecision(5)) - offset[1],
    ]
  }

  static getPressure(e: PointerEvent | React.PointerEvent | Touch | React.Touch | WheelEvent) {
    return 'pressure' in e ? Number(e.pressure.toPrecision(5)) || 0.5 : 0.5
  }

  static commandKey(): string {
    return Utils.isDarwin() ? 'âŒ˜' : 'Ctrl'
  }
}

export const inputs = new Inputs()
