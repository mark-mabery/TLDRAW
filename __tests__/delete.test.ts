import state from 'state'
import inputs from 'state/inputs'
import { ShapeType } from 'types'
import { getShape } from 'utils'
import {
  idsAreSelected,
  point,
  rectangleId,
  arrowId,
  getOnlySelectedShape,
  assertShapeProps,
} from './test-utils'
import * as json from './__mocks__/document.json'

describe('deleting single shapes', () => {
  state.reset()
  state.send('MOUNTED').send('LOADED_FROM_FILE', { json: JSON.stringify(json) })

  it('deletes a shape and undoes the delete', () => {
    state
      .send('CANCELED')
      .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
      .send('STOPPED_POINTING', inputs.pointerUp(point(), rectangleId))

    expect(idsAreSelected(state.data, [rectangleId])).toBe(true)

    state.send('DELETED')

    expect(idsAreSelected(state.data, [])).toBe(true)
    expect(getShape(state.data, rectangleId)).toBe(undefined)

    state.send('UNDO')

    expect(getShape(state.data, rectangleId)).toBeTruthy()
    expect(idsAreSelected(state.data, [rectangleId])).toBe(true)

    state.send('REDO')

    expect(getShape(state.data, rectangleId)).toBe(undefined)

    state.send('UNDO')
  })
})

describe('deletes and restores grouped shapes', () => {
  state.reset()
  state.send('MOUNTED').send('LOADED_FROM_FILE', { json: JSON.stringify(json) })

  it('creates a group', () => {
    state
      .send('CANCELED')
      .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
      .send('STOPPED_POINTING', inputs.pointerUp(point(), rectangleId))
      .send(
        'POINTED_SHAPE',
        inputs.pointerDown(point({ shiftKey: true }), arrowId)
      )
      .send(
        'STOPPED_POINTING',
        inputs.pointerUp(point({ shiftKey: true }), arrowId)
      )

    expect(idsAreSelected(state.data, [rectangleId, arrowId])).toBe(true)

    state.send('GROUPED')

    const group = getOnlySelectedShape(state.data)

    // Should select the group
    expect(assertShapeProps(group, { type: ShapeType.Group }))

    const arrow = tld.getShape(state.data, arrowId)

    // The arrow should be have the group as its parent
    expect(assertShapeProps(arrow, { parentId: group.id }))
  })

  // it('selects the new group', () => {
  //   expect(idsAreSelected(state.data, [groupId])).toBe(true)
  // })

  // it('assigns a new parent', () => {
  //   expect(groupId === state.data.currentPageId).toBe(false)
  // })

  // // Rectangle has the same new parent?
  // it('assigns new parent to all selected shapes', () => {
  //   expect(hasParent(state.data, arrowId, groupId)).toBe(true)
  // })

  // // New parent is selected?
  // it('selects the new parent', () => {
  //   expect(idsAreSelected(state.data, [groupId])).toBe(true)
  // })
})

//   // it('selects the group when pointing a shape', () => {
//   //   state
//   //     .send('CANCELED')
//   //     .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
//   //     .send('STOPPED_POINTING', inputs.pointerUp(point(), rectangleId))

//   //   expect(idsAreSelected(state.data, [groupId])).toBe(true)
//   // })

//   // it('keeps selection when pointing bounds', () => {
//   //   state
//   //     .send('CANCELED')
//   //     .send('POINTED_BOUNDS', inputs.pointerDown(point(), 'bounds'))
//   //     .send('STOPPED_POINTING', inputs.pointerUp(point(), 'bounds'))

//   //   expect(idsAreSelected(state.data, [groupId])).toBe(true)
//   // })

//   // it('selects a grouped shape by double-pointing', () => {
//   //   state
//   //     .send('CANCELED')
//   //     .send('DOUBLE_POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
//   //     .send('STOPPED_POINTING', inputs.pointerUp(point(), rectangleId))

//   //   expect(idsAreSelected(state.data, [rectangleId])).toBe(true)
//   // })

//   // it('selects a sibling on point when selecting a grouped shape', () => {
//   //   state
//   //     .send('POINTED_SHAPE', inputs.pointerDown(point(), arrowId))
//   //     .send('STOPPED_POINTING', inputs.pointerUp(point(), arrowId))

//   //   expect(idsAreSelected(state.data, [arrowId])).toBe(true)
//   // })

//   // it('rises up a selection level when escape is pressed', () => {
//   //   state
//   //     .send('CANCELED')
//   //     .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
//   //     .send('STOPPED_POINTING', inputs.pointerUp(point(), rectangleId))

//   //   expect(idsAreSelected(state.data, [groupId])).toBe(true)
//   // })

//   // it('deletes and restores one shape', () => {
//   //   // Delete the rectangle first
//   //   state.send('UNDO')

//   //   expect(getShape(state.data, rectangleId)).toBeTruthy()
//   //   expect(idsAreSelected(state.data, [rectangleId])).toBe(true)

//   //   state.send('REDO')

//   //   expect(getShape(state.data, rectangleId)).toBe(undefined)

//   //   state.send('UNDO')

//   //   expect(getShape(state.data, rectangleId)).toBeTruthy()
//   //   expect(idsAreSelected(state.data, [rectangleId])).toBe(true)
//   // })
// })
