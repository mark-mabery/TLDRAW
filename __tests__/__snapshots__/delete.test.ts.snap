// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deletes and restores grouped shapes creates a group: data after mount from file 1`] = `
Object {
  "code": Object {
    "file0": Object {
      "code": "",
      "id": "file0",
      "name": "index.ts",
    },
  },
  "id": "0001",
  "name": "My Document",
  "pages": Object {
    "page1": Object {
      "childIndex": 0,
      "id": "page1",
      "name": "Page 1",
      "shapes": Object {
        "1f6c251c-e12e-40b4-8dd2-c1847d80b72f": Object {
          "childIndex": 24,
          "id": "1f6c251c-e12e-40b4-8dd2-c1847d80b72f",
          "isAspectRatioLocked": false,
          "isGenerated": false,
          "isHidden": false,
          "isLocked": false,
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            0,
            0,
          ],
          "radius": 2,
          "rotation": 0,
          "seed": 0.6440313303074272,
          "size": Array [
            67.22075383450237,
            72.92795609221832,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Solid",
            "isFilled": false,
            "size": "Small",
          },
          "type": "rectangle",
        },
        "5ca167d7-54de-47c9-aa8f-86affa25e44d": Object {
          "bend": 0,
          "childIndex": 16,
          "decorations": Object {
            "end": null,
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                3.2518097616315345,
                140.54510317291172,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                0,
                0,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                6.503619523263069,
                281.09020634582345,
              ],
            },
          },
          "id": "5ca167d7-54de-47c9-aa8f-86affa25e44d",
          "isAspectRatioLocked": false,
          "isGenerated": false,
          "isHidden": false,
          "isLocked": false,
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            100,
            100,
          ],
          "points": Array [
            Array [
              6.503619523263069,
              281.09020634582345,
            ],
            Array [
              0,
              0,
            ],
          ],
          "rotation": 0,
          "seed": 0.08116783083496548,
          "style": Object {
            "color": "Black",
            "dash": "Solid",
            "isFilled": false,
            "size": "Small",
          },
          "type": "arrow",
        },
      },
      "type": "page",
    },
  },
}
`;
